File: tool/csv2json.py
import pandas as pd
import json
from argparse import ArgumentParser

SIRE_RANK = [
    # ランキング上位
    "キズナ",
    "エピファネイア",
    "スワーヴリチャード",
    "モーリス",
    "リアルスティール",
    "ディープインパクト",
    "ロードカナロア",
    "ドレフォン",
    "ブリックスアンドモルタル",
    "ニューイヤーズデイ",
    "ドゥラメンテ",
    "ダイワメジャー",
    "レイデオロ",
    "ルーラーシップ",
    "ヘニーヒューズ",
    "ゴールドシップ",
    # その他
    "ハービンジャー",
    "ハーツクライ",
    "キタサンブラック",
    "シルバーステート",
    "オルフェーヴル"
    # 新種牡馬
    "ルヴァンスレーヴ",
    "ゴールドドリーム",
    "サートゥルナーリア",
    "ウインブライト",
    "アドマイヤマーズ"
]

parser = ArgumentParser()
parser.add_argument("csvfile", type=str, help="csvfile")
parser.add_argument("-o", "--output", type=str, default="horse_catalogue.json", help="output filename")
args = parser.parse_args()

df = pd.read_csv(args.csvfile)
missing_rows = df["name"].isnull() | df["sire"].isnull() | df["mare"].isnull()
for _, row in df[missing_rows].iterrows():
    if isinstance(row["name"], float):
        print(f"Missing name: {row['id']}")
    if isinstance(row["sire"], float):
        print(f"Missing sire: {row['id']}")
    if isinstance(row["mare"], float):
        print(f"Missing mare: {row['id']}")
df = df[~missing_rows]
df['sort_order'] = df['sire'].apply(lambda x: len(SIRE_RANK) - SIRE_RANK.index(x) if x in SIRE_RANK else 0)
df = df.sort_values('sort_order', ascending=False)
data = []
for i, row in df.iterrows():
    data.append({"id": row["id"], "name": row["name"], "sire": row["sire"], "mare": row["mare"]})
json.dump(data, open(args.output, "w"), ensure_ascii=False)

File: tool/csv2json.js
const fs = require('fs');
const csv = require('csv-parser');
const yargs = require('yargs/yargs');
const { hideBin } = require('yargs/helpers');


const argv = yargs(hideBin(process.argv))
    .option('output', {
        alias: 'o',
        type: 'string',
        default: 'horse_catalogue.json',
        description: 'output filename'
    })
    .demandCommand(1)
    .help()
    .argv;

const csvfile = argv._[0];
const outputFile = argv.output;

const results = [];
const missingRows = [];

fs.createReadStream(csvfile)
    .pipe(csv())
    .on('data', (row) => {
        results.push(row);
    })
    .on('end', () => {
        // 欠損値のチェック
        const validRows = results.filter(row => {
            const isValid = row.name && row.sire && row.mare;
            if (!isValid) {
                if (!row.name) {
                    console.log(`Missing name: ${row.id}`);
                }
                if (!row.sire) {
                    console.log(`Missing sire: ${row.id}`);
                }
                if (!row.mare) {
                    console.log(`Missing mare: ${row.id}`);
                }
                missingRows.push(row);
            }
            return isValid;
        });

        // 種牡馬ごとの出現回数をカウント
        const sireCounts = {};
        validRows.forEach(row => {
            if (row.sire) {
                sireCounts[row.sire] = (sireCounts[row.sire] || 0) + 1;
            }
        });

        // 種牡馬の出現回数を各行に追加
        const rowsWithSireCount = validRows.map(row => ({
            ...row,
            sire_count: sireCounts[row.sire] || 0
        }));

        // 種牡馬の出現回数で降順ソート
        rowsWithSireCount.sort((a, b) => b.sire_count - a.sire_count);

        // 必要なデータだけを抽出
        const data = rowsWithSireCount.map(row => ({
            id: row.id,
            name: row.name,
            sire: row.sire,
            mare: row.mare,
            sire_count: row.sire_count
        }));

        // JSONファイルに書き込み
        fs.writeFileSync(outputFile, JSON.stringify(data, null, 2));
        console.log(`Converted data saved to ${outputFile}`);
    });
